python/ray/dashboard/dashboard.py

```diff
@@ -0,0 +1,325 @@
+
+    def run(self):
+        self.log_dashboard_url()
+        self.node_stats.start()
+        aiohttp.web.run_app(self.app, host=self.ip, port=self.port)
+
+
+class NodeStats(threading.Thread):
+    def __init__(self, redis_address, redis_password=None):
+        self.redis_key = "{}.*".format(ray.gcs_utils.REPORTER_CHANNEL)
+        self.redis_client = ray.services.create_redis_client(
+            redis_address, password=redis_password)
+
+        self._node_stats = {}
+        self._node_stats_lock = threading.Lock()
+        super().__init__()
+
+    def calculate_totals(self) -> Dict:
+        total_boot_time = 0
+        total_cpus = 0
+        total_workers = 0
+        total_load = [0.0,0.0,0.0]
+        total_storage_avail = 0
+        total_storage_total = 0
+        total_ram_avail = 0
+        total_ram_total = 0
+        total_sent = 0
+        total_recv = 0
+
+        for v in self._node_stats.values():
+            total_boot_time += v["boot_time"]
+            total_cpus += v["cpus"][0]
+            total_workers += len(v["workers"])
+            total_load[0] += v["load_avg"][0][0]
+            total_load[1] += v["load_avg"][0][1]
+            total_load[2] += v["load_avg"][0][2]
+            total_storage_avail += v["disk"]["/"]["free"]
+            total_storage_total += v["disk"]["/"]["total"]
+            total_ram_avail += v["mem"][1]
+            total_ram_total += v["mem"][0]
+            total_sent += v["net"][0]
+            total_recv += v["net"][1]
+
+        return {
+            "boot_time": total_boot_time,
+            "n_workers": total_workers,
+            "n_cores": total_cpus,
+            "m_avail": total_ram_avail,
+            "m_total": total_ram_total,
+            "d_avail": total_storage_avail,
+            "d_total": total_storage_total,
+            "load": total_load,
+            "n_sent": total_sent,
+            "n_recv": total_recv,
+        }
+
+    def calculate_tasks(self) -> Counter:
+        return Counter(
+            (x["name"] for y in (v["workers"] for v in self._node_stats.values()) for x in y)
+        )
+
+    def purge_outdated_stats(self):
+        def current(then, now):
+            if (now - then) > 5:
+                return False
+
+            return True
+
+        now = to_unix_time(datetime.datetime.utcnow())
+        self._node_stats = {
+            k: v
+            for k, v in self._node_stats.items()
+            if current(v["now"], now)
+        }
+
+    def get_node_stats(self) -> Dict:
+        with self._node_stats_lock:
+            self.purge_outdated_stats()
+            node_stats = sorted(
+                (v for v in self._node_stats.values()),
+                key=itemgetter("boot_time")
+            )
+            return {
+                "totals": self.calculate_totals(),
+                "tasks": self.calculate_tasks(),
+                "clients": node_stats,
+            }
+
+    def run(self):
+        p = self.redis_client.pubsub()
```

Comment:
You can do `pubsub(ignore_subscribe_messages=True)` to avoid one of the cases below.

Reply:
Done